pkg/scheduler/eventhandlers/reservation_handler.go

```diff
@@ -0,0 +1,150 @@
+			oldR, oldOK := oldObj.(*schedulingv1alpha1.Reservation)
+			newR, newOK := newObj.(*schedulingv1alpha1.Reservation)
+			if !oldOK || !newOK {
+				klog.Errorf("unable to convert object to *schedulingv1alpha1.Reservation, old %T, new %T", oldObj, newObj)
+				return
+			}
+			updateReservation(sched, oldR, newR)
+		},
+		DeleteFunc: func(obj interface{}) {
+			switch t := obj.(type) {
+			case *schedulingv1alpha1.Reservation:
+				deleteReservation(sched, t)
+				return
+			case cache.DeletedFinalStateUnknown:
+				r, ok := t.Obj.(*schedulingv1alpha1.Reservation)
+				if ok {
+					deleteReservation(sched, r)
+					return
+				}
+			}
+			klog.Errorf("reservation informer delete func parse obj failed, obj %T", obj)
+		},
+	})
+}
+
+func addReservation(sched *scheduler.Scheduler, r *schedulingv1alpha1.Reservation) {
+	if r == nil || r.Spec.Template == nil {
+		klog.Errorf("failed to add reservation for invalid reservation %v", r)
+		return
+	}
+	reservePod := reservation.NewReservePod(r)
+	// TBD: should the reservation with nodeName specified enqueue?
+	if r.Status.Phase != schedulingv1alpha1.ReasonReservationExpired {
+		err := sched.SchedulingQueue.Add(reservePod)
+		if err == nil {
+			klog.V(4).InfoS("add reservation, enqueue reserve pod to scheduling queue",
+				"reservation", reservation.GetReservationNamespacedName(r))
+		} else {
+			klog.Errorf("failed to add reserve pod into scheduling queue, reservation %v, err: %v",
+				reservation.GetReservationNamespacedName(r), err)
+		}
+	}
+}
+
+func updateReservation(sched *scheduler.Scheduler, oldR, newR *schedulingv1alpha1.Reservation) {
+	if oldR == nil || newR == nil {
+		klog.Errorf("failed to update reservation for invalid reservation old %v, new %v", oldR, newR)
+		return
+	}
+	// only react to status.phase change
+	// NOTE: currently changing reservation spec is not supported, and the new spec will be ignored.
+	if oldR.Status.Phase == newR.Status.Phase {
+		klog.V(5).InfoS("skip for reservation update since its status.phase is not change",
+			"reservation", reservation.GetReservationNamespacedName(newR))
+		return
+	}
+	klog.V(3).InfoS("got reservation phase update",
+		"reservation", reservation.GetReservationNamespacedName(newR), "old", oldR.Status.Phase, "new", newR.Status.Phase)
+
+	if oldR.Status.Phase == schedulingv1alpha1.ReservationAvailable &&
+		newR.Status.Phase == schedulingv1alpha1.ReservationExpired {
+		deleteReservation(sched, newR)
+	}
+}
+
+func deleteReservation(sched *scheduler.Scheduler, r *schedulingv1alpha1.Reservation) {
+	if r == nil || r.Spec.Template == nil {
+		klog.Errorf("failed to delete reservation for invalid reservation %v", r)
+		return
+	}
+
+	reservePod := reservation.NewReservePod(r)
+	// if reservation is scheduled, delete it from the schedule cache;
+	// otherwise, remove it from the scheduling queue.
+
+	if reservation.IsReservationScheduled(r) {
+		err := sched.SchedulerCache.RemovePod(reservePod)
```

Comment:
We also need to be careful with Assume Pods. The current process also lacks this.